src/core/interpreter/profile-parameter-validator.test.ts

```diff
@@ -22,1005 +20,768 @@ const checkErrorKind = (result: Result<unknown, unknown>) =>
-      });
-
-      it('should fail with invalid input', () => {
-        const result1 = parameterValidator.validate(
-          { test: 'none of your business' },
-          'input',
-          'Test'
-        );
-        expect(checkErrorKind(result1)).toEqual(['enumValue']);
-        expect(checkErrorPath(result1)).toEqual([['input', 'test']]);
-        const result2 = parameterValidator.validate(
-          { test: 7 },
-          'input',
-          'Test'
-        );
-        expect(checkErrorKind(result2)).toEqual(['enumValue']);
-        expect(checkErrorPath(result2)).toEqual([['input', 'test']]);
+      it('returns error for invalid values', () => {
+        expect(parameterValidator.validate({ foo: 1 }, 'result', 'Test').isErr()).toBeTruthy();
       });
     });
 
-    describe('AST with field-referenced named enum', () => {
+    describe('union', () => {
       const ast = parseProfileFromSource(`
-        usecase Test safe {
-          input {
-            test
+        usecase Test {
+          result {
+            foo boolean | number | string
+            bar Foo | Bar
+            baz boolean | Foo | Baz | [string]
           }
         }
 
-        model TestEnum enum { A, B, C = 'CC', D }
-        field test TestEnum
+        model Foo {
+          foo string
+        }
+        model Bar {
+          foo number
+          bar string
+        }
+        model Baz enum { OK, ERR }
       `);
-      let parameterValidator: ProfileParameterValidator;
 
       beforeEach(() => {
         parameterValidator = new ProfileParameterValidator(ast);
       });
 
-      it.each(['A', 'B', 'CC', 'D', null])(
-        'should pass with valid input: %s',
-        value => {
-          expect(
-            parameterValidator.validate({ test: value }, 'input', 'Test').isOk()
-          ).toEqual(true);
-        }
-      );
+      it('returns ok for valid primitives in foo', () => {
+        expect(parameterValidator.validate({ foo: true, }, 'result', 'Test').isOk()).toBeTruthy();
+        expect(parameterValidator.validate({ foo: 1, }, 'result', 'Test').isOk()).toBeTruthy();
+        expect(parameterValidator.validate({ foo: 'value', }, 'result', 'Test').isOk()).toBeTruthy();
+      });
 
-      it.each(['a', 7, true])('should fail with invalid value: %p', value => {
-        const result = parameterValidator.validate(
-          { test: value },
-          'input',
-          'Test'
-        );
-        expect(checkErrorKind(result)).toEqual(['enumValue']);
-        expect(checkErrorPath(result)).toEqual([['input', 'test']]);
+      it('returns ok for valid object in bar', () => {
+        expect(parameterValidator.validate({ bar: { foo: 'value' } }, 'result', 'Test').isOk()).toBeTruthy();
+        expect(parameterValidator.validate({ bar: { foo: 1, bar: 'value' } }, 'result', 'Test').isOk()).toBeTruthy();
+        expect(parameterValidator.validate({ bar: { foo: 'value', bar: 'value' } }, 'result', 'Test').isOk()).toBeTruthy();
```

Comment:
Maybe you are right, but I don't think the change you are suggesting should be part of this PR.

Reply:
I agree, just trying to analyze what I think we should take away from these new, shiny, and readable tests.

Reply:
So, what is your suggestion here? Remove the test?

Reply:
My suggestion would be t invert the test and treat it like a bug that needs fixing in context of this feature. We could also stick with it and say we want it as a behavior.

Reply:
You mean in this PR or in this feature (None type)?

Reply:
Ok, I moved the test and marked it should result to error.

Reply:
I had a second (or maybe third) look at this. And realized that the reason why `{ foo: 'value', bar: 'value' }` passes validation is that it is valid `Foo` because we allow extraneous fields. I will compare how json-schema validation works and some up with proposal how it should behave.

```
usecase Test {
  result {
    foo boolean | number | string
    bar Foo | Bar
    baz boolean | Foo | Baz | [string]
  }
}

model Foo {
  foo string
}
model Bar {
  foo number
  bar string
}
model Baz enum { OK, ERR }
```